exec: handle NULL group by keys in hash aggregator

rafiss opened the pull request:
Previously, if an aggregation was requested with NULL group keys, an
index out of bounds error would occur in the hash aggregator. The reason
was that the hashTable data structure is shared between the hash joiner
and the hash aggregator. It was originally implemented for the hash joiner,
and in that case, NULLs are always supposed to be treated as non-equal.
However, in the hash aggregator, we want NULLs to compare as equal to each
other.

This change allows the tests in logic_test/aggregate to pass with the
local-vec configuration.

In order to exercise more of the null handling logic, the tests now set
garbage data in the corresponing position of the vector whenever an
element is null.

fixes #38750 

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38900)
<!-- Reviewable:end -->

yuzefovich reviewed (changes requested):
Nice investigation! Your explanation makes sense to me, but I think a minor additional adjustment is needed for this PR to be bullet-proof.

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38900#-:-LjvneRh6rhzBHleuTYS:b1ukk9m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/exec/hashjoiner_tmpl.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38900#-Ljvn2RQ9U1jPvesX8RI:-Ljvn2RR9Z_Ul4AmFLiN:b-9jkhdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/36a9bcb410f03ca32eba4025f841ec905288e3a3/pkg/sql/exec/hashjoiner_tmpl.go#L73)):*
> ```Go
> 	probeVal := probeKeys[_SEL_IND]
> 	var unique bool
> 	_ASSIGN_NE(unique, buildVal, probeVal)
> ```

I think this part needs an adjustment. In theory, we can have two Vecs with nulls at the appropriate indices but with garbage in the actual values. Null bitmap takes a priority, so it would still be a correct representation. But we would get that two nulls are different because of the leftover garbage.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38900)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/exec/hashjoiner_tmpl.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38900#-Ljvn2RQ9U1jPvesX8RI:-LjvqCtg4qpDmAdv6joz:b-128sue) ([raw file](https://github.com/cockroachdb/cockroach/blob/36a9bcb410f03ca32eba4025f841ec905288e3a3/pkg/sql/exec/hashjoiner_tmpl.go#L73)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think this part needs an adjustment. In theory, we can have two Vecs with nulls at the appropriate indices but with garbage in the actual values. Null bitmap takes a priority, so it would still be a correct representation. But we would get that two nulls are different because of the leftover garbage.
</blockquote></details>

thanks, this makes sense... hmm, i wonder then, i think i also will need to adjust the `rehash` function in hashjoiner_tmpl.go so that it computes a hash consistently when something is null.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38900)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/exec/hashjoiner_tmpl.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38900#-Ljvn2RQ9U1jPvesX8RI:-LjvrkMX7blUNO7L7pEO:bw9ncz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/36a9bcb410f03ca32eba4025f841ec905288e3a3/pkg/sql/exec/hashjoiner_tmpl.go#L73)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

thanks, this makes sense... hmm, i wonder then, i think i also will need to adjust the `rehash` function in hashjoiner_tmpl.go so that it computes a hash consistently when something is null.
</blockquote></details>

Good point. I agree, `rehash` needs an adjustment as well.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38900)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/exec/hashjoiner_tmpl.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38900#-Ljvn2RQ9U1jPvesX8RI:-LjvrwcxCKJXgwHusgur:bw0w85) ([raw file](https://github.com/cockroachdb/cockroach/blob/36a9bcb410f03ca32eba4025f841ec905288e3a3/pkg/sql/exec/hashjoiner_tmpl.go#L73)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Good point. I agree, `rehash` needs an adjustment as well.
</blockquote></details>

I think we should make somewhat artificial test that would trigger these possible issues with nulls and with garbage in values. Maybe other places need adjustments, I'm not too familiar with this code to be honest.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38900)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/exec/hashjoiner_tmpl.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38900#-Ljvn2RQ9U1jPvesX8RI:-LjwyJpn0sA-P8MexmFn:b-whhlud) ([raw file](https://github.com/cockroachdb/cockroach/blob/36a9bcb410f03ca32eba4025f841ec905288e3a3/pkg/sql/exec/hashjoiner_tmpl.go#L73)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think we should make somewhat artificial test that would trigger these possible issues with nulls and with garbage in values. Maybe other places need adjustments, I'm not too familiar with this code to be honest.
</blockquote></details>

Hmm, I might have an idea for a different approach. Could we edit `_CHECK_COL_BODY` to have a 4th case, for when ProbeHasNulls is true, BuildHasNulls is true, and also the allowNullEquality boolean (that could be a template variable, btw) is also true? Then, `_CHECK_COL_BODY` could have special case handling for the situation in which both sides are NULL. I think?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38900)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rafiss)

___
*[pkg/sql/exec/hashjoiner_tmpl.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38900#-Ljvn2RQ9U1jPvesX8RI:-Ljx-hO2EtjHHPTTWDA0:b-5yl2nm) ([raw file](https://github.com/cockroachdb/cockroach/blob/36a9bcb410f03ca32eba4025f841ec905288e3a3/pkg/sql/exec/hashjoiner_tmpl.go#L73)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Hmm, I might have an idea for a different approach. Could we edit `_CHECK_COL_BODY` to have a 4th case, for when ProbeHasNulls is true, BuildHasNulls is true, and also the allowNullEquality boolean (that could be a template variable, btw) is also true? Then, `_CHECK_COL_BODY` could have special case handling for the situation in which both sides are NULL. I think?
</blockquote></details>

I feel like `rehash` will still need to be taught about `allowNulEquality`.


<!-- Sent from Reviewable.io -->

rafiss committed:
exec: handle NULL group by keys in hash aggregator

Previously, if an aggregation was requested with NULL group keys, an
index out of bounds error would occur in the hash aggregator. The reason
was that the hashTable data structure is shared between the hash joiner
and the hash aggregator. It was originally implemented for the hash joiner,
and in that case, NULLs are always supposed to be treated as non-equal.
However, in the hash aggregator, we want NULLs to compare as equal to each
other.

This change allows the tests in logic_test/aggregate to pass with the
local-vec configuration.

In order to exercise more of the null handling logic, the tests now set
garbage data in the corresponing position of the vector whenever an
element is null.

Release note: None

rafiss reviewed (commented):
please take another look!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38900#-:-Lk01aHK3gICoMpap3_k:b-vwkhtv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/exec/hashjoiner_tmpl.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38900#-Ljvn2RQ9U1jPvesX8RI:-Lk01J8QDiKT1w31oNuB:b7rptr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/36a9bcb410f03ca32eba4025f841ec905288e3a3/pkg/sql/exec/hashjoiner_tmpl.go#L73)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I feel like `rehash` will still need to be taught about `allowNulEquality`.
</blockquote></details>

Sorry, I didn't notice this comment thread. I think I ended up doing basically as Jordan suggested, but maybe it's a little more clunky than what you had in mind. I also did fix `rehash`, and also edited test infrastructure so garbage is inserted whenever something is set to NULL.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Thanks! I cannot say that I fully understand how vectorized hash table works, but this change <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38900#-:-Lk04KD54w816o0rs2TU:b-1gvd4u)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @rafiss)

___
*[pkg/sql/exec/utils_test.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38900#-Lk03qvu3wmqfHdUAwC7:-Lk03qvu3wmqfHdUAwC8:b3uji3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/0890944363e2ef886d87cdd2ba727c67d3a6ff7d/pkg/sql/exec/utils_test.go#L300)):*
> ```Go
> 					col.Index(int(outputIdx)).Set(val)
> 				} else {
> 					panic(fmt.Sprintf("Could not generate a random value pf type %s\n.", typ.Name()))
> ```

[nit]: `s/Could/could/` and `s/pf/of/`.


<!-- Sent from Reviewable.io -->

rafiss commented:
benchmark numbers aren't affected too much

```
name                                                                   old time/op    new time/op    delta
HashJoiner/nulls=false/fullOuter=false/distinct=true/rows=2048-24         562µs ± 4%     567µs ± 7%    ~     (p=0.436 n=10+10)
HashJoiner/nulls=false/fullOuter=false/distinct=true/rows=262144-24      44.0ms ± 2%    43.0ms ± 2%  -2.30%  (p=0.000 n=10+10)
HashJoiner/nulls=false/fullOuter=false/distinct=true/rows=4194304-24      668ms ± 1%     661ms ± 2%  -1.16%  (p=0.002 n=8+9)
HashJoiner/nulls=false/fullOuter=false/distinct=false/rows=2048-24        545µs ± 1%     548µs ± 2%    ~     (p=0.173 n=8+10)
HashJoiner/nulls=false/fullOuter=false/distinct=false/rows=262144-24     43.9ms ± 2%    42.7ms ± 4%  -2.59%  (p=0.002 n=10+10)
HashJoiner/nulls=false/fullOuter=false/distinct=false/rows=4194304-24     1.21s ± 7%     1.15s ± 4%  -5.29%  (p=0.034 n=10+8)
HashJoiner/nulls=false/fullOuter=true/distinct=true/rows=2048-24          583µs ± 1%     561µs ± 4%  -3.71%  (p=0.000 n=9+10)
HashJoiner/nulls=false/fullOuter=true/distinct=true/rows=262144-24       45.1ms ± 1%    44.3ms ± 1%  -1.81%  (p=0.000 n=10+10)
HashJoiner/nulls=false/fullOuter=true/distinct=true/rows=4194304-24       690ms ± 2%     679ms ± 2%  -1.61%  (p=0.002 n=10+9)
HashJoiner/nulls=false/fullOuter=true/distinct=false/rows=2048-24         558µs ± 2%     530µs ± 9%  -5.07%  (p=0.001 n=10+10)
HashJoiner/nulls=false/fullOuter=true/distinct=false/rows=262144-24      43.4ms ± 3%    41.5ms ± 2%  -4.42%  (p=0.000 n=10+10)
HashJoiner/nulls=false/fullOuter=true/distinct=false/rows=4194304-24      1.00s ± 5%     1.02s ± 5%    ~     (p=0.280 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=true/rows=2048-24          643µs ± 3%     662µs ± 1%  +2.93%  (p=0.000 n=10+9)
HashJoiner/nulls=true/fullOuter=false/distinct=true/rows=262144-24       53.2ms ± 1%    54.2ms ± 1%  +2.00%  (p=0.000 n=9+10)
HashJoiner/nulls=true/fullOuter=false/distinct=true/rows=4194304-24       811ms ± 1%     827ms ± 1%  +2.01%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=false/rows=2048-24         624µs ± 4%     633µs ± 3%    ~     (p=0.190 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=false/rows=262144-24      57.2ms ± 2%    54.9ms ± 1%  -3.97%  (p=0.000 n=9+9)
HashJoiner/nulls=true/fullOuter=false/distinct=false/rows=4194304-24      1.38s ± 4%     1.39s ± 6%    ~     (p=0.529 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=true/rows=2048-24           656µs ± 1%     665µs ± 1%  +1.35%  (p=0.002 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=true/rows=262144-24        54.4ms ± 1%    56.1ms ± 3%  +3.18%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=true/rows=4194304-24        824ms ± 1%     844ms ± 1%  +2.41%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=false/rows=2048-24          625µs ± 2%     646µs ± 1%  +3.32%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=false/rows=262144-24       53.6ms ± 2%    53.0ms ± 1%  -1.09%  (p=0.043 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=false/rows=4194304-24       1.14s ± 3%     1.17s ± 3%    ~     (p=0.089 n=10+10)

name                                                                   old speed      new speed      delta
HashJoiner/nulls=false/fullOuter=false/distinct=true/rows=2048-24       233MB/s ± 4%   232MB/s ± 7%    ~     (p=0.436 n=10+10)
HashJoiner/nulls=false/fullOuter=false/distinct=true/rows=262144-24     381MB/s ± 2%   390MB/s ± 2%  +2.35%  (p=0.000 n=10+10)
HashJoiner/nulls=false/fullOuter=false/distinct=true/rows=4194304-24    402MB/s ± 1%   406MB/s ± 2%  +1.18%  (p=0.002 n=8+9)
HashJoiner/nulls=false/fullOuter=false/distinct=false/rows=2048-24      240MB/s ± 1%   239MB/s ± 2%    ~     (p=0.173 n=8+10)
HashJoiner/nulls=false/fullOuter=false/distinct=false/rows=262144-24    382MB/s ± 2%   393MB/s ± 3%  +2.69%  (p=0.002 n=10+10)
HashJoiner/nulls=false/fullOuter=false/distinct=false/rows=4194304-24   221MB/s ± 8%   233MB/s ± 4%  +5.42%  (p=0.034 n=10+8)
HashJoiner/nulls=false/fullOuter=true/distinct=true/rows=2048-24        225MB/s ± 1%   234MB/s ± 4%  +3.91%  (p=0.000 n=9+10)
HashJoiner/nulls=false/fullOuter=true/distinct=true/rows=262144-24      372MB/s ± 1%   379MB/s ± 1%  +1.84%  (p=0.000 n=10+10)
HashJoiner/nulls=false/fullOuter=true/distinct=true/rows=4194304-24     389MB/s ± 2%   396MB/s ± 2%  +1.63%  (p=0.002 n=10+9)
HashJoiner/nulls=false/fullOuter=true/distinct=false/rows=2048-24       235MB/s ± 2%   248MB/s ± 9%  +5.51%  (p=0.001 n=10+10)
HashJoiner/nulls=false/fullOuter=true/distinct=false/rows=262144-24     387MB/s ± 3%   405MB/s ± 2%  +4.61%  (p=0.000 n=10+10)
HashJoiner/nulls=false/fullOuter=true/distinct=false/rows=4194304-24    269MB/s ± 5%   264MB/s ± 5%    ~     (p=0.280 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=true/rows=2048-24        204MB/s ± 3%   198MB/s ± 1%  -2.86%  (p=0.000 n=10+9)
HashJoiner/nulls=true/fullOuter=false/distinct=true/rows=262144-24      316MB/s ± 1%   309MB/s ± 1%  -1.96%  (p=0.000 n=9+10)
HashJoiner/nulls=true/fullOuter=false/distinct=true/rows=4194304-24     331MB/s ± 1%   324MB/s ± 1%  -1.97%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=false/rows=2048-24       210MB/s ± 4%   207MB/s ± 3%    ~     (p=0.190 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=false/rows=262144-24     293MB/s ± 2%   306MB/s ± 1%  +4.13%  (p=0.000 n=9+9)
HashJoiner/nulls=true/fullOuter=false/distinct=false/rows=4194304-24    195MB/s ± 4%   193MB/s ± 6%    ~     (p=0.529 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=true/rows=2048-24         200MB/s ± 1%   197MB/s ± 1%  -1.34%  (p=0.002 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=true/rows=262144-24       309MB/s ± 1%   299MB/s ± 3%  -3.07%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=true/rows=4194304-24      326MB/s ± 1%   318MB/s ± 1%  -2.36%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=false/rows=2048-24        210MB/s ± 2%   203MB/s ± 1%  -3.22%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=false/rows=262144-24      313MB/s ± 2%   316MB/s ± 1%  +1.10%  (p=0.043 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=false/rows=4194304-24     235MB/s ± 3%   230MB/s ± 3%    ~     (p=0.089 n=10+10)

name                                                                   old alloc/op   new alloc/op   delta
HashJoiner/nulls=false/fullOuter=false/distinct=true/rows=2048-24         805kB ± 0%     805kB ± 0%  +0.00%  (p=0.000 n=9+8)
HashJoiner/nulls=false/fullOuter=false/distinct=true/rows=262144-24      46.4MB ± 0%    46.4MB ± 0%  +0.00%  (p=0.000 n=9+10)
HashJoiner/nulls=false/fullOuter=false/distinct=true/rows=4194304-24      697MB ± 0%     697MB ± 0%  +0.00%  (p=0.001 n=8+10)
HashJoiner/nulls=false/fullOuter=false/distinct=false/rows=2048-24        809kB ± 0%     809kB ± 0%  +0.00%  (p=0.000 n=10+8)
HashJoiner/nulls=false/fullOuter=false/distinct=false/rows=262144-24     44.6MB ± 0%    44.6MB ± 0%  +0.00%  (p=0.001 n=10+9)
HashJoiner/nulls=false/fullOuter=false/distinct=false/rows=4194304-24     668MB ± 0%     668MB ± 0%    ~     (p=0.222 n=10+10)
HashJoiner/nulls=false/fullOuter=true/distinct=true/rows=2048-24          808kB ± 0%     808kB ± 0%  +0.00%  (p=0.000 n=10+9)
HashJoiner/nulls=false/fullOuter=true/distinct=true/rows=262144-24       46.7MB ± 0%    46.7MB ± 0%  +0.00%  (p=0.003 n=10+10)
HashJoiner/nulls=false/fullOuter=true/distinct=true/rows=4194304-24       701MB ± 0%     701MB ± 0%  +0.00%  (p=0.000 n=9+10)
HashJoiner/nulls=false/fullOuter=true/distinct=false/rows=2048-24         812kB ± 0%     812kB ± 0%  +0.00%  (p=0.000 n=10+7)
HashJoiner/nulls=false/fullOuter=true/distinct=false/rows=262144-24      44.9MB ± 0%    44.9MB ± 0%  +0.00%  (p=0.000 n=10+10)
HashJoiner/nulls=false/fullOuter=true/distinct=false/rows=4194304-24      672MB ± 0%     672MB ± 0%    ~     (p=0.338 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=true/rows=2048-24          805kB ± 0%     805kB ± 0%  -0.00%  (p=0.000 n=10+9)
HashJoiner/nulls=true/fullOuter=false/distinct=true/rows=262144-24       46.4MB ± 0%    46.4MB ± 0%  -0.02%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=true/rows=4194304-24       697MB ± 0%     697MB ± 0%  -0.02%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=false/rows=2048-24         809kB ± 0%     809kB ± 0%  -0.00%  (p=0.000 n=9+10)
HashJoiner/nulls=true/fullOuter=false/distinct=false/rows=262144-24      44.6MB ± 0%    44.6MB ± 0%  -0.01%  (p=0.000 n=9+9)
HashJoiner/nulls=true/fullOuter=false/distinct=false/rows=4194304-24      668MB ± 0%     668MB ± 0%  -0.01%  (p=0.000 n=6+10)
HashJoiner/nulls=true/fullOuter=true/distinct=true/rows=2048-24           808kB ± 0%     808kB ± 0%  -0.00%  (p=0.000 n=9+8)
HashJoiner/nulls=true/fullOuter=true/distinct=true/rows=262144-24        46.7MB ± 0%    46.7MB ± 0%  -0.02%  (p=0.000 n=8+10)
HashJoiner/nulls=true/fullOuter=true/distinct=true/rows=4194304-24        701MB ± 0%     701MB ± 0%  -0.02%  (p=0.000 n=10+8)
HashJoiner/nulls=true/fullOuter=true/distinct=false/rows=2048-24          812kB ± 0%     812kB ± 0%  -0.00%  (p=0.000 n=9+6)
HashJoiner/nulls=true/fullOuter=true/distinct=false/rows=262144-24       44.9MB ± 0%    44.9MB ± 0%  -0.01%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=false/rows=4194304-24       672MB ± 0%     672MB ± 0%  -0.01%  (p=0.000 n=9+10)

name                                                                   old allocs/op  new allocs/op  delta
HashJoiner/nulls=false/fullOuter=false/distinct=true/rows=2048-24         8.28k ± 0%     8.28k ± 0%    ~     (all equal)
HashJoiner/nulls=false/fullOuter=false/distinct=true/rows=262144-24       1.05M ± 0%     1.05M ± 0%    ~     (all equal)
HashJoiner/nulls=false/fullOuter=false/distinct=true/rows=4194304-24      16.8M ± 0%     16.8M ± 0%    ~     (all equal)
HashJoiner/nulls=false/fullOuter=false/distinct=false/rows=2048-24        6.24k ± 0%     6.24k ± 0%    ~     (all equal)
HashJoiner/nulls=false/fullOuter=false/distinct=false/rows=262144-24       527k ± 0%      527k ± 0%    ~     (all equal)
HashJoiner/nulls=false/fullOuter=false/distinct=false/rows=4194304-24     8.40M ± 0%     8.40M ± 0%    ~     (p=0.721 n=10+10)
HashJoiner/nulls=false/fullOuter=true/distinct=true/rows=2048-24          8.28k ± 0%     8.28k ± 0%    ~     (all equal)
HashJoiner/nulls=false/fullOuter=true/distinct=true/rows=262144-24        1.05M ± 0%     1.05M ± 0%    ~     (all equal)
HashJoiner/nulls=false/fullOuter=true/distinct=true/rows=4194304-24       16.8M ± 0%     16.8M ± 0%    ~     (p=0.248 n=9+10)
HashJoiner/nulls=false/fullOuter=true/distinct=false/rows=2048-24         6.24k ± 0%     6.24k ± 0%    ~     (all equal)
HashJoiner/nulls=false/fullOuter=true/distinct=false/rows=262144-24        527k ± 0%      527k ± 0%    ~     (all equal)
HashJoiner/nulls=false/fullOuter=true/distinct=false/rows=4194304-24      8.40M ± 0%     8.40M ± 0%    ~     (p=0.173 n=10+9)
HashJoiner/nulls=true/fullOuter=false/distinct=true/rows=2048-24          8.28k ± 0%     8.27k ± 0%  -0.10%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=true/rows=262144-24        1.05M ± 0%     1.05M ± 0%  -0.10%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=true/rows=4194304-24       16.8M ± 0%     16.8M ± 0%  -0.10%  (p=0.001 n=8+9)
HashJoiner/nulls=true/fullOuter=false/distinct=false/rows=2048-24         6.24k ± 0%     6.23k ± 0%  -0.10%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=false/rows=262144-24        527k ± 0%      527k ± 0%  -0.10%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=false/distinct=false/rows=4194304-24      8.41M ± 0%     8.40M ± 0%  -0.10%  (p=0.000 n=7+10)
HashJoiner/nulls=true/fullOuter=true/distinct=true/rows=2048-24           8.28k ± 0%     8.28k ± 0%  -0.10%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=true/rows=262144-24         1.05M ± 0%     1.05M ± 0%  -0.10%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=true/rows=4194304-24        16.8M ± 0%     16.8M ± 0%  -0.10%  (p=0.002 n=8+10)
HashJoiner/nulls=true/fullOuter=true/distinct=false/rows=2048-24          6.24k ± 0%     6.23k ± 0%  -0.10%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=false/rows=262144-24         527k ± 0%      527k ± 0%  -0.10%  (p=0.000 n=10+10)
HashJoiner/nulls=true/fullOuter=true/distinct=false/rows=4194304-24       8.40M ± 0%     8.40M ± 0%  -0.10%  (p=0.000 n=9+10)
```

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38900)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/exec/utils_test.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38900#-Lk03qvu3wmqfHdUAwC7:-Lk0AgeQCXeN8_0yCZt1:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/0890944363e2ef886d87cdd2ba727c67d3a6ff7d/pkg/sql/exec/utils_test.go#L300)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: `s/Could/could/` and `s/pf/of/`.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

rafiss commented:
thanks all
bors r+

rafiss commented:
bors r+
